### ISSUE TITLE
twisted.trial.runner.TrialRunner._removeSafely has typo and untested branch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4020]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#4020|
|Type|defect|
|Created|2009-09-14 05:56:30Z|
|Branch|https://github.com/twisted/twisted/tree/remove-safely-4020|

It incorrectly uses `FilePath`, which doesn't exist. (`filepath.FilePath` does)

http://twistedmatrix.com/trac/browser/trunk/twisted/trial/runner.py#L728

This was found with Pyflakes.

<details><summary>Searchable metadata</summary>

```
trac-id__4020 4020
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__branches_remove_safely_4020 branches/remove-safely-4020
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1252907790000000 1252907790000000
changetime__1263169971000000 1263169971000000
version__None None
owner__jesstess jesstess
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#4135](https://github.com/twisted/twisted/issues/4135) and [#4193](https://github.com/twisted/twisted/issues/4193) were duplicates of this.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

(In [27904]) Branching to 'remove-safely-4020'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess commented|
|-|-|

(In [27905]) Fix typo and add unit tests for the 3 possible branches in
_removeSafely.

refs [#4020](https://github.com/twisted/twisted/issues/4020)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess removed owner|
|-|-|

Changes to the branch include:
 * fixing the typo
 * epytext in _removeSafely
 * addition of unit tests for each of the 3 branches in _removeSafely
 * updating the copyrights on runner.py and test_runner.py
 * addition of .misc file

Buildbot results [here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-safely-4020).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

I'd say it's a bug so a bugfix news file might be more appropriate..

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [27909]) .misc ==> .bugfix

refs [#4020](https://github.com/twisted/twisted/issues/4020)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|TimAllen set owner to @jesstess|
|-|-|

1.If I were reading a hypothetical Twisted 10.0 announcement, I'd find it difficult to figure out what the message in your `.bugfix` file meant to me as an ordinary developer. Something like "Twisted Trial no longer crashes if it can't remove an old `_trial_temp` directory" would be more explanatory, I think. People can look up the ticket number if they want the embarrassing details.
 1. In `twisted/trial/test/test_runner.py`, you might as well define `dummyRemove()` and `dummyMoveTo()` inside the tests where they're used; they're pretty simple.
 1. Again in the tests, rather than "`filepath.FilePath(os.path.join(directory, '_trial_marker')).touch()`" you could just say "`dirPath.child('_trial_marker').touch()`"
 1. Why do you call `stdout.restore()` before the final `assertEquals`? `self.patch()` monkey-patches are automatically reverted at the end of the test anyway.
 1. Rather than slicing the output string for comparison purposes, just say `failUnlessIn("could not remove FilePath", out.getvalue())`.
 1. In `test_removeSafelyRemoveFailsMoveFails`, you should check that the `OSError` that gets re-raised is the one you expect - make `dummyRemove()` and `dummyMoveTo()` pass some recognizable parameter to `OSError()` and check the properties of the `OSError` that `failUnlessRaises` returns. Likewise, you should use `failUnlessIn` to check that both error messages are present in the captured stdout.
 1. Why have you added these tests to `twisted.trial.test.test_runner.TestUntilFailure`? You could have added them to `twisted.trial.test.test_runner.TestRunner` and replaced the higher-level `test_noMarker` test in that `TestCase`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Coding standard note, _For consistency, Twisted unit tests should use the assert forms rather than the fail forms and the singular names rather than the plural names._

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess commented|
|-|-|

(In [27939]) Move trial marker tests into test_runner.TestRunner and address the
rest of TimAllen's review points.

refs [#4020](https://github.com/twisted/twisted/issues/4020)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess removed owner|
|-|-|

 * re 1-6: done.
 * re 7: done. No reason other than 'I didn't see test_noMarker when I was glancing through the file to decide where to put these tests'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. Avoid the `fail*` assertions.  `assertRaises` should be used instead of `failUnlessRaises`.  `assertIn` should be used instead of `failUnlessIn`.
  1. Use the ends-with-s spellings of the `assert*` methods.  `assertEquals` instead of `assertEqual`, etc.
  1. What are the new pairs of single quotes in the _removeSafely docstring around `C{_trial_marker}`?  `C{"_trial_marker"}` would make sense, I think.

Otherwise looks good.  Please merge once the above points are fixed.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess set status to `closed`|
|-|-|

(In [27950]) Merge remove-safely-4020

Author: jesstess
Reviewers: TimAllen, exarkun
Fixes: [#4020](https://github.com/twisted/twisted/issues/4020)

Fix namespace typo so that Twisted Trial no longer crashes if it can't
remove an old _trial_temp directory, and add unit tests for the
previously untested branches in TrialRunner._removeSafely.
